src/marzullo.zig

```diff
@@ -0,0 +1,306 @@
+    pub const Interval = struct {
+        /// The lower bound on the minimum clock offset.
+        lower_bound: i64,
+
+        /// The upper bound on the maximum clock offset.
+        upper_bound: i64,
+
+        /// The number of "true chimers" consistent with the largest number of sources.
+        sources_true: u8,
+
+        /// The number of "false chimers" falling outside this interval.
+        /// Where `sources_false` plus `sources_true` always equals the total number of sources.
+        sources_false: u8,
+    };
+
+    /// A tuple represents either the lower or upper end of a bound, and is fed as input to the
+    /// Marzullo algorithm to compute the smallest interval across all tuples.
+    /// For example, given a clock offset to a remote replica of 3s, a round trip time of 1s, and
+    /// a maximum tolerance between clocks of 100ms on either side, we might create two tuples, the
+    /// lower bound having an offset of 2.4s and the upper bound having an offset of 3.6s,
+    /// to represent the error introduced by the round trip time and by the clocks themselves.
+    pub const Tuple = struct {
+        source: u8,
+        offset: i64,
+        bound: enum {
+            lower,
+            upper,
+        },
+    };
+
+    /// Returns the smallest interval consistent with the largest number of sources.
+    pub fn smallest_interval(tuples: []Tuple) Interval {
+        // There are two bounds (lower and upper) per clock offset sample.
+        const bounds = 2;
+
+        const sources = @divExact(tuples.len, bounds);
+        assert(sources <= std.math.maxInt(u8));
+
+        if (sources == 0) {
+            return Interval{
+                .lower_bound = 0,
+                .upper_bound = 0,
+                .sources_true = 0,
+                .sources_false = 0,
+            };
+        }
+
+        std.sort.insertionSort(Tuple, tuples, {}, less_than);
+
+        // Here is a description of the algorithm:
+        // https://en.wikipedia.org/wiki/Marzullo%27s_algorithm#Method
+        var best: i64 = 0;
+        var count: i64 = 0;
+        var previous: ?Tuple = null;
+        var interval: Interval = undefined;
+
+        for (tuples) |tuple, i| {
+            // Verify that our sort implementation is correct:
+            if (previous) |p| {
+                assert(p.offset <= tuple.offset);
+                if (p.offset == tuple.offset) {
+                    if (p.bound != tuple.bound) {
+                        assert(p.bound == .lower and tuple.bound == .upper);
+                    } else {
+                        assert(p.source < tuple.source);
+                    }
+                }
+            }
+            previous = tuple;
+
+            // Update the current number of overlapping intervals:
+            count -= switch (tuple.bound) {
+                .lower => @as(i64, -1),
+                .upper => @as(i64, 1),
+            };
```

Comment:
I find the `-= -1` here not very readable, and the `@as()` is pretty noisy (though only needed due to a limitation of the stage1 compilner). I'd write
```zig
            switch (tuple.bound) {
                .lower => count += 1,
                .upper => count -= 1,
            }
```

Reply:
Much better, thanks! The `-= -1` of the algorithm confused me at first, so I tried to show the bounds with enum literals. Getting away from the `-=` entirely is great.